File: apps/android/core/src/main/kotlin/ai/deepseek/dsh/companion/LiteFold.kt
package ai.deepseek.dsh.companion

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

/** Why a Lite turn ended. */
typealias LiteTurnEnd = String

/** One completed conversation row. */
data class LiteMessage(val role: String, val text: String, val interrupted: Boolean = false)

/** One todo row; the wire status rides verbatim. */
data class LiteTodo(val content: String, val status: String)

/** One tool invocation, paired by id. */
data class LiteToolCall(
    val id: String,
    val name: String,
    val arguments: String,
    val phase: String,
    val resultText: String,
)

/** One artifact reference — metadata only; content never rides the spec (chapter 56). */
data class LiteArtifact(val id: String, val kind: String, val title: String, val status: String)

/** One recorded failure. */
data class LiteFailure(val kind: String, val code: String, val message: String)

/** The live-stream pane: whether a turn is streaming and its delivered partials. */
data class LiteStreaming(val active: Boolean = false, val partialText: String = "", val partialReasoning: String = "")

/**
 * The complete Lite-visible runtime state at one event cut — the Kotlin half
 * of the Lite Behavior-Spec conformance: the same lite-conformance fixtures
 * the TypeScript reference fold and the Swift Lite fold replay must fold to
 * this exact state here.
 */
data class LiteDomainState(
    val conversation: List<LiteMessage> = emptyList(),
    val streaming: LiteStreaming = LiteStreaming(),
    val interrupted: Boolean = false,
    val toolCalls: List<LiteToolCall> = emptyList(),
    val planActive: Boolean = false,
    val todos: List<LiteTodo> = emptyList(),
    val artifacts: List<LiteArtifact> = emptyList(),
    val lastTurnEnd: LiteTurnEnd? = null,
    val errors: List<LiteFailure> = emptyList(),
    val pendingHandoff: String? = null,
)

/** The state before any event arrives. */
fun emptyLiteDomain(): LiteDomainState = LiteDomainState()

/**
 * Fold a complete Lite event sequence into the runtime state — the exact
 * mirror of the TypeScript reference fold and the Swift Lite fold: a cancel
 * finalizes the delivered stream prefix as an interrupted assistant row, a
 * dropped transport keeps the prefix for resume, and whole-value panes are
 * last-write-wins.
 * @param events ordered Lite lifecycle events as raw JSON objects.
 * @returns the derived domain state.
 */
fun foldLiteDomain(events: JsonArray): LiteDomainState {
    val state = LiteFoldAccumulator()
    for (element in events) {
        val event = element as? JsonObject ?: continue
        foldLiteEventInto(state, event)
    }
    return state.toLiteDomainState()
}

/** Mutable accumulator the fold loop mutates in place. */
private class LiteFoldAccumulator {
    val conversation = mutableListOf<LiteMessage>()
    var streaming = LiteStreaming()
    var interrupted = false
    val toolCalls = mutableListOf<LiteToolCall>()
    var planActive = false
    var todos: List<LiteTodo> = emptyList()
    val artifacts = mutableListOf<LiteArtifact>()
    var lastTurnEnd: LiteTurnEnd? = null
    val errors = mutableListOf<LiteFailure>()
    var pendingHandoff: String? = null
}

private fun LiteFoldAccumulator.toLiteDomainState() = LiteDomainState(
    conversation = conversation.toList(),
    streaming = streaming,
    interrupted = interrupted,
    toolCalls = toolCalls.toList(),
    planActive = planActive,
    todos = todos,
    artifacts = artifacts.toList(),
    lastTurnEnd = lastTurnEnd,
    errors = errors.toList(),
    pendingHandoff = pendingHandoff,
)

private fun JsonObject.stringField(name: String): String? =
    (this[name] as? JsonPrimitive)?.takeIf { it.isString }?.content

private fun JsonObject.booleanField(name: String): Boolean? =
    this[name]?.jsonPrimitive?.booleanOrNull

/**
 * Fold one Lite lifecycle event into the runtime state. Unknown tags are
 * no-ops; the events arrive from the pinned Lite fixtures and journals, so
 * the vocabulary is closed.
 */
private fun foldLiteEventInto(state: LiteFoldAccumulator, event: JsonObject) {
    when (event.stringField("type")) {
        "prompt/accepted" -> state.conversation.add(LiteMessage(role = "user", text = event.stringField("content") ?: ""))
        "prompt/rejected" -> state.errors.add(
            LiteFailure(kind = "provider", code = "PROMPT_REJECTED", message = event.stringField("reason") ?: ""),
        )
        "stream/delta" -> state.streaming = state.streaming.copy(
            active = true,
            partialText = state.streaming.partialText + (event.stringField("text") ?: ""),
        )
        "stream/reasoning" -> state.streaming = state.streaming.copy(
            active = true,
            partialReasoning = state.streaming.partialReasoning + (event.stringField("text") ?: ""),
        )
        "message/completed" -> {
            state.conversation.add(LiteMessage(role = "assistant", text = event.stringField("text") ?: ""))
            state.streaming = LiteStreaming()
        }
        "turn/completed" -> state.lastTurnEnd = "completed"
        "turn/cancelled" -> {
            if (state.streaming.active && state.streaming.partialText.isNotEmpty()) {
                state.conversation.add(LiteMessage(role = "assistant", text = state.streaming.partialText, interrupted = true))
                state.interrupted = true
            }
            state.streaming = LiteStreaming()
            state.lastTurnEnd = "cancelled"
        }
        "tool/call" -> state.toolCalls.add(
            LiteToolCall(
                id = event.stringField("id") ?: "",
                name = event.stringField("name") ?: "",
                arguments = event.stringField("arguments") ?: "",
                phase = "running",
                resultText = "",
            ),
        )
        "tool/result" -> {
            val id = event.stringField("id")
            val index = if (id == null) -1 else state.toolCalls.indexOfFirst { it.id == id }
            if (index != -1) {
                val target = state.toolCalls[index]
                state.toolCalls[index] = target.copy(
                    phase = if (event.booleanField("ok") == true) "completed" else "failed",
                    resultText = event.stringField("text") ?: "",
                )
            }
        }
        "plan/changed" -> state.planActive = event.booleanField("active") == true
        "todo/changed" -> state.todos = (event["todos"] as? JsonArray ?: JsonArray(emptyList())).mapNotNull { entry ->
            val obj = entry as? JsonObject ?: return@mapNotNull null
            LiteTodo(
                content = obj.stringField("content") ?: "",
                status = obj.stringField("status") ?: "",
            )
        }
        "artifact/created" -> state.artifacts.add(
            LiteArtifact(
                id = event.stringField("id") ?: "",
                kind = event.stringField("kind") ?: "",
                title = event.stringField("title") ?: "",
                status = "pending",
            ),
        )
        "artifact/status" -> {
            val id = event.stringField("id")
            val index = if (id == null) -1 else state.artifacts.indexOfFirst { it.id == id }
            if (index != -1) {
                state.artifacts[index] = state.artifacts[index].copy(status = event.stringField("status") ?: "")
            }
        }
        "provider/error" -> {
            state.errors.add(LiteFailure(kind = "provider", code = event.stringField("code") ?: "", message = event.stringField("message") ?: ""))
            state.streaming = LiteStreaming()
            state.lastTurnEnd = "provider-error"
        }
        "network/error" -> {
            // A dropped transport keeps the delivered prefix for resume; the
            // stream is no longer live.
            state.streaming = state.streaming.copy(active = false)
            val kind = event.stringField("kind") ?: ""
            state.errors.add(LiteFailure(kind = "network", code = kind, message = kind))
            state.lastTurnEnd = "network-error"
        }
        "handoff/requested" -> state.pendingHandoff = event.stringField("capability")
    }
}

/**
 * Canonical JSON form of the state — an interrupted marker only on rows it
 * is true for, booleans and nullable ends verbatim — so conformance compares
 * against the expected fixture bytes structurally.
 */
fun LiteDomainState.toJson(): JsonElement = buildJsonObject {
    put("conversation", JsonArray(conversation.map { message ->
        buildJsonObject {
            put("role", message.role)
            put("text", message.text)
            if (message.interrupted) put("interrupted", true)
        }
    }))
    put("streaming", buildJsonObject {
        put("active", streaming.active)
        put("partialText", streaming.partialText)
        put("partialReasoning", streaming.partialReasoning)
    })
    put("interrupted", interrupted)
    put("toolCalls", JsonArray(toolCalls.map { call ->
        buildJsonObject {
            put("id", call.id)
            put("name", call.name)
            put("arguments", call.arguments)
            put("phase", call.phase)
            put("resultText", call.resultText)
        }
    }))
    put("planActive", planActive)
    put("todos", JsonArray(todos.map { todo ->
        buildJsonObject {
            put("content", todo.content)
            put("status", todo.status)
        }
    }))
    put("artifacts", JsonArray(artifacts.map { artifact ->
        buildJsonObject {
            put("id", artifact.id)
            put("kind", artifact.kind)
            put("title", artifact.title)
            put("status", artifact.status)
        }
    }))
    put("lastTurnEnd", JsonPrimitive(lastTurnEnd))
    put("errors", JsonArray(errors.map { failure ->
        buildJsonObject {
            put("kind", failure.kind)
            put("code", failure.code)
            put("message", failure.message)
        }
    }))
    put("pendingHandoff", JsonPrimitive(pendingHandoff))
}

/** Parse one lite-conformance scenario document. */
fun parseLiteScenario(text: String): Pair<JsonArray, JsonObject> {
    val document = kotlinx.serialization.json.Json.parseToJsonElement(text).jsonObject
    val events = document["events"]?.jsonArray ?: JsonArray(emptyList())
    val expected = document["expected"]?.jsonObject ?: JsonObject(emptyMap())
    return Pair(events, expected)
}

File: apps/apple/Sources/LiteRuntime/LiteFold.swift
import Foundation

/// One completed conversation row.
public struct LiteMessage: Equatable, Decodable {
    public let role: String
    public let text: String
    public let interrupted: Bool?

    public init(role: String, text: String, interrupted: Bool? = nil) {
        self.role = role
        self.text = text
        self.interrupted = interrupted
    }
}

/// One paired tool invocation.
public struct LiteToolRecord: Equatable, Decodable {
    public enum Phase: String, Equatable, Decodable {
        case running, completed, failed
    }

    public let id: String
    public let name: String
    public let arguments: String
    public let phase: Phase
    public let resultText: String

    public init(id: String, name: String, arguments: String, phase: Phase, resultText: String) {
        self.id = id
        self.name = name
        self.arguments = arguments
        self.phase = phase
        self.resultText = resultText
    }
}

/// One artifact reference — metadata only, never content.
public struct LiteArtifactRecord: Equatable, Decodable {
    public let id: String
    public let kind: String
    public let title: String
    public let status: LiteArtifactStatus

    public init(id: String, kind: String, title: String, status: LiteArtifactStatus) {
        self.id = id
        self.kind = kind
        self.title = title
        self.status = status
    }
}

/// One recorded failure.
public struct LiteFailureRecord: Equatable, Decodable {
    public let kind: String
    public let code: String
    public let message: String
}

/// Why a Lite turn ended.
public enum LiteTurnEnd: String, Equatable, Decodable {
    case completed
    case cancelled
    case providerError = "provider-error"
    case networkError = "network-error"
}

/// The complete Lite-visible runtime state at one event cut — JSON-keyed to
/// the TypeScript reference fold's emission.
public struct LiteDomainState: Equatable, Decodable {
    public struct Streaming: Equatable, Decodable {
        public var active: Bool
        public var partialText: String
        public var partialReasoning: String

        public init(active: Bool = false, partialText: String = "", partialReasoning: String = "") {
            self.active = active
            self.partialText = partialText
            self.partialReasoning = partialReasoning
        }
    }

    public var conversation: [LiteMessage]
    public var streaming: Streaming
    public var interrupted: Bool
    public var toolCalls: [LiteToolRecord]
    public var planActive: Bool
    public var todos: [LiteTodo]
    public var artifacts: [LiteArtifactRecord]
    public var lastTurnEnd: LiteTurnEnd?
    public var errors: [LiteFailureRecord]
    public var pendingHandoff: String?

    public init(
        conversation: [LiteMessage] = [], streaming: Streaming = Streaming(active: false, partialText: "", partialReasoning: ""),
        interrupted: Bool = false, toolCalls: [LiteToolRecord] = [], planActive: Bool = false,
        todos: [LiteTodo] = [], artifacts: [LiteArtifactRecord] = [], lastTurnEnd: LiteTurnEnd? = nil,
        errors: [LiteFailureRecord] = [], pendingHandoff: String? = nil
    ) {
        self.conversation = conversation
        self.streaming = streaming
        self.interrupted = interrupted
        self.toolCalls = toolCalls
        self.planActive = planActive
        self.todos = todos
        self.artifacts = artifacts
        self.lastTurnEnd = lastTurnEnd
        self.errors = errors
        self.pendingHandoff = pendingHandoff
    }
}

/// The Swift half of the Lite Behavior Spec: folds lifecycle events into
/// the domain state with exactly the TypeScript reference's semantics — a
/// cancel finalizes the delivered stream prefix as an interrupted assistant
/// row, a network drop stops the stream but keeps the partial, a provider
/// error clears streaming and sets the terminal outcome.
public struct LiteFold {
    public private(set) var state: LiteDomainState

    public init() {
        state = LiteDomainState()
    }

    /// Fold one lifecycle event.
    public mutating func apply(_ event: LiteEvent) {
        switch event {
        case .promptAccepted(_, let content):
            state.conversation.append(LiteMessage(role: "user", text: content))
        case .promptRejected(_, let reason):
            state.errors.append(LiteFailureRecord(kind: "provider", code: "PROMPT_REJECTED", message: reason))
        case .streamDelta(let text):
            state.streaming.active = true
            state.streaming.partialText += text
        case .streamReasoning(let text):
            state.streaming.active = true
            state.streaming.partialReasoning += text
        case .messageCompleted(let text, _):
            state.conversation.append(LiteMessage(role: "assistant", text: text))
            state.streaming = .init(active: false, partialText: "", partialReasoning: "")
        case .turnCompleted:
            state.lastTurnEnd = .completed
        case .turnCancelled:
            if state.streaming.active && !state.streaming.partialText.isEmpty {
                state.conversation.append(LiteMessage(role: "assistant", text: state.streaming.partialText, interrupted: true))
                state.interrupted = true
            }
            state.streaming = .init(active: false, partialText: "", partialReasoning: "")
            state.lastTurnEnd = .cancelled
        case .toolCall(let id, let name, let arguments):
            state.toolCalls.append(LiteToolRecord(id: id, name: name, arguments: arguments, phase: .running, resultText: ""))
        case .toolResult(let id, let ok, let text):
            guard let index = state.toolCalls.firstIndex(where: { $0.id == id }) else { return }
            let call = state.toolCalls[index]
            state.toolCalls[index] = LiteToolRecord(
                id: call.id, name: call.name, arguments: call.arguments,
                phase: ok ? .completed : .failed, resultText: text
            )
        case .planChanged(let active):
            state.planActive = active
        case .todoChanged(let todos):
            state.todos = todos
        case .artifactCreated(let id, let kind, let title):
            state.artifacts.append(LiteArtifactRecord(id: id, kind: kind, title: title, status: .pending))
        case .artifactStatus(let id, let status):
            if let index = state.artifacts.firstIndex(where: { $0.id == id }) {
                let record = state.artifacts[index]
                state.artifacts[index] = LiteArtifactRecord(id: record.id, kind: record.kind, title: record.title, status: status)
            }
        case .providerError(let code, let message):
            state.errors.append(LiteFailureRecord(kind: "provider", code: code, message: message))
            state.streaming = .init(active: false, partialText: "", partialReasoning: "")
            state.lastTurnEnd = .providerError
        case .networkError(let kind):
            state.streaming.active = false
            state.errors.append(LiteFailureRecord(kind: "network", code: kind, message: kind))
            state.lastTurnEnd = .networkError
        case .handoffRequested(let capability):
            state.pendingHandoff = capability
        }
    }
}
